aeppl/truncation.py

```diff
@@ -129,3 +144,216 @@ def censor_logprob(op, values, base_rv, lower_bound, upper_bound, **kwargs):
         )
 
     return logprob
+
+
+class TruncatedRV(OpFromGraph):
+    """An `Op` constructed from an Aesara graph that represents a truncated univariate RV."""
+
+    default_output = 1
+    base_rv_op = None
+
+    def __init__(self, base_rv_op: Op, *args, **kwargs):
+        self.base_rv_op = base_rv_op
+        super().__init__(*args, **kwargs)
+
+
+@singledispatch
+def _truncated(op: Op, lower, upper, *params):
+    """Return the truncated equivalent of another ``RandomVariable``."""
+    raise NotImplementedError(
+        f"{op} does not have an equivalent truncated version implemented"
+    )
+
+
+def truncate(
+    rv: TensorVariable, lower=None, upper=None, max_n_steps: int = 10_000, rng=None
+):
+    """Truncate a univariate RandomVariable between lower and upper.
+
+    If lower or upper is ``None``, the variable is not truncated on that side.
+
+    Depending on dispatched implementations, this function returns either a specialized
+    `Op`, or equivalent graph representing the truncation process, via inverse CDF
+    sampling, or rejection sampling.
+
+    The argument `max_n_steps` controls the maximum number of resamples that are
+    attempted when performing rejection sampling. An Error is raised if convergence is
+    not reached after that many steps.
+
+    TODO: Add Note about updates
+    """
+
+    if lower is None and upper is None:
+        raise ValueError("lower and upper cannot both be None")
+
+    if not (isinstance(rv.owner.op, RandomVariable) and rv.owner.op.ndim_supp == 0):
+        raise NotImplementedError(
+            f"truncation only implemented for univariate random variables, got {rv.owner.op}"
+        )
+
+    lower = at.as_tensor_variable(lower) if lower is not None else at.constant(-np.inf)
+    upper = at.as_tensor_variable(upper) if upper is not None else at.constant(np.inf)
+
+    if rng is None:
+        rng = shared(np.random.RandomState(), borrow=True)
```

Comment:
We need to use `RandomStream`s instead.

Reply:
In this case I think `RandomStream` causes more trouble than it helps. We can end up creating several intermediate RVs by the time we return the final graph, and `RandomStream` would be creating a new RNG shared variable and chaining them for each intermediate object, even if they are not returned. The only thing I really need is one RandomState variable

Reply:
> In this case I think `RandomStream` causes more trouble than it helps. We can end up creating several intermediate RVs by the time we return the final graph, and `RandomStream` would be creating a new RNG shared variable and chaining them for each intermediate object, even if they are not returned. The only thing I really need is one RandomState variable

`RandomStream` is the standard user interface for generating random samples in Aesara, so it **must** work with it.

If you believe there are issues with `RandomStream` itself, then those need to be addressed separately; otherwise, the things you mentioned are not issues that justify not using `RandomStream`.  

For instance, if you're concerned about the number of RNG objects created, then remember that `RandomStream` creates the shared RNG objects (i.e. they are not created in/by the compiled graph), so once `RandomStream.some_method` is called by the user, the work is already done.  If you're concerned that the user-level use of `RandomStream` is slow for this reason, then that would be surprising, because creating NumPy RNG objects should take no noticeable time.

The only in-graph concern is whether or not the `RandomVariable` `Op`s have been in-lined, but using `RandomStream` would make no difference in this case.

More importantly, you need to consider the **consistency** of the approach you're advocating.  Use of `RandomStream` guarantees that the sample results will be consistent with their user-level seeding, whereas the current approach does not.  

For instance, if the `rng` object in this code is used by two distinct but equivalent `RandomVariable`s and neither are in-lined, they will both produce the same sample values, and, if they are in-lined, they will produce distinct samples, and that isn't consistent.

There's more to it, but these reasons alone are enough to warrant the use of the `RandomStream` interface.

Reply:
I guess my issue with `RandomStream` is that it tries to do too much. It assumes what I want as updates and it introduces default_updates on the rng, which are a mess for any further graph manipulations and have been a constant source of bugs on my end.

Reply:
> I guess my issue with `RandomStream` is that it tries to do too much. It assumes what I want as updates and it introduces default_updates on the rng, which are a mess for any further graph manipulations and have been a constant source of bugs on my end.

Unless you faithfully reproduce what `RandomStream` does, the resulting graphs aren't guaranteed to be numerically consistent and/or they'll be considerably more difficult to optimize (e.g. due to the introduction of unnecessary subgraph dependencies); otherwise, if you reproduce what `RandomStream` already does, then it's just going to make the implementation more complicated/verbose.

Reply:
`RandomStream` does 3 things from what I gathered:
1. Create new SharedRandomStates from an initial seed
2. Add the updates of a new RV to a list and set them as default_update of the respective SharedRandomState
3. Reseed all RandomStates

My issue is only with step 2, and concretely with setting the `default_update`. I used to think it was necessary for use in Scan's but that does not seem to be the case. We just need to return updates explicitly instead of relying on the fallback behavior where it intercepts the `default_updates`.

The reason I don't like default_updates or hard-coded updates is that it complicates the manipulation of random graphs. I think we should be setting them prior to function compilation...

Reply:
> My issue is only with step 2. I used to think the `default_update` was necessary for use in Scan's but that does not seem to be the case. We just need to return updates explicitly instead of relying on the fallback behavior where it intercepts the `default_updates`.

As I stated above, the use of shared variables and `default_update` produces **independent** RNG state objects, and this is important for maintaining the applicability of optimizations.  Also, there currently isn't a numerically consistent non-shared variable/updates-based approach for updating an RNG state in `Scan` or `OpFromGraph`&mdash;unless `RandomVariable.inplace == True`, which isn't the case by default and breaks the underlying purity/reproducibility of our graphs.  

See https://github.com/aesara-devs/aesara/issues/738, https://github.com/aesara-devs/aesara/issues/898, and https://github.com/aesara-devs/aesara/issues/454 for more details about the aforementioned points.

> The reason I don't like default_updates or a frozen list of updates is that it complicates the manipulation of random graphs. Every time I want to work with a `Random Op` or an `OpFramGraph` that works with `SharedRandomStates`, I will have to worry about whether such variables came from a RandomStream or not. Even if I change the `default_update` (and this is making an assumption that the user did not put it there on purpose), there will now be an inconsistency between the `default_update` and the one stored in `RandomStream.updates()`.

Shared variables and `default_update` do not complicate anything that lies within the established framework for graph construction and manipulation; however, if you're trying to do something that **isn't** supported, then, yes, you'll need to account for things on your own.  `OpFromGraph` is a good example of something that **explicitly does not support updates** and, as a result, it does not work with `default_update`s or `Scan`s (that generate updates).  **This is an `OpFromGraph`-specific issue**, and it can only be solved by updating `OpFromGraph`.

Regardless, the alternative non-shared approaches I've seen are simply inconsistent, because they reuse the same RNG state(s) across random variables that are supposed to be entirely distinct, so we cannot use them.  Likewise, a user could add `default_update`s and shared variables of their own (i.e. outside of `RandomStream`), and **our code should still work**, so there's no avoiding the need to correctly handle those features.


> As a clear example, this aeppl helper is currently not safe to use with RVs with default updates, and was a source of quite bad bug in PyMC:
> 
> https://github.com/aesara-devs/aeppl/blob/87416830d7439cb64aba4d508361b14012227b2e/aeppl/utils.py#L68

That helper function (i.e. `change_rv_size`) was originally a quick hack to get something working; however, it was not a viable approach so we stopped using it (and it needs to be removed from this library).  The correct approach to that functionality is an `Op` that represents a change in the size of a `RandomVariable` (e.g. some sort of "replicate" `Op`).

Nevertheless, `change_rv_size` isn't used and shouldn't be, and not because of `RandomStream`, shared variables, or `default_update`, but because it produces a graph-level change in a `Variable` that occurs outside of the standard graph manipulation framework, and it has numerous fundamental issues due to **that** alone (e.g. any maps from/to the original variable will be broken once the results from that function are used in place of the original variable).

> The problem I have with RandomStream is that it makes it quite hard to manipulate graphs with RVs for the user convenience. I am fine with behaviors 1 and 3.

Again, it seems like the only place that actual problems appear are the places where updates are not immediately supported, which is currently **only** `OpFromGraph`.  Aside from that, I'm not aware of any other situations in which updates or `default_update` complicate graph manipulation.  

Regarding `Scan`, the answer is simple: return any updates explicitly created in a `Scan` loop function.  The only time that's really relevant is when a `Scan` loop function creates a `Scan`, and, in that scenario, the updates are already a return value that you're forced to deal with, so there are no surprises involved.  Again, this is something one must do regardless of `RandomVariable`s, `RandomStream`s, etc., because updates can be introduced in other ways.

Furthermore, now that https://github.com/aesara-devs/aesara/pull/910 is merged, the interaction between `Scan` and `default_update`s is more consistent.  If there are any other issues, we simply need to fix them like we did with that one, but we cannot introduce other inconsistencies or complicated code and graph structures in an attempt to avoid a narrow subset of problems that have already been resolved and/or have very straightforward solutions.